pkg/controller/flinkapplication/flink_state_machine.go

```diff
@@ -227,6 +228,12 @@ func (s *FlinkStateMachine) handleNewOrUpdating(ctx context.Context, application
 		return applicationUnchanged, err
 	}
 
+	// If there are old resources left-over from a previous version, clean them up
+	err = s.flinkController.DeleteOldResourcesForApp(ctx, application)
+	if err != nil {
+		logger.Warn(ctx, "Failed to clean up old resources: %v", err)
+	}
+
```

Comment:
These several lines take care of deleting the old cluster.

Reply:
@yuchaoran2011 This does not look right. 

The cluster is deleted in the Running state. 

The sequence is Running - Updating to Savepointing to ClusterStarting - SubmittingJob and back to Running in the new cluster. 

Only after the job is running in the new cluster the old cluster is deleted. This is done in handleApplicationRunning. Let me know if you are noticing something else

Reply:
That's not what I observed though. What I saw was that when the application is running, after I apply some changes to the CR, then a new cluster is started. The application then starts running in the new cluster, but the old cluster is still around/not deleted. You should be able to reproduce it by doing what I described.

Reply:
We have integration tests to verify. Wondering if we missed an edge case. Can you paste the status of the application resource, and deployments?

This is the code for deletion - https://github.com/lyft/flinkk8soperator/blob/master/pkg/controller/flinkapplication/flink_state_machine.go#L545

Reply:
I see what you meant now. You got a valid point. Let me check with my colleague who reported the issue to me. If his application did not successfully savepoint, then the old cluster not being deleted by the operator is expected. Only if his application did successfully savepoint, then is it an operator problem. I'll get back to you when I hear from him.

Reply:
I just got some information from my colleague @debasishg on this issue. Here's the pods available after he scaled the FlinkApp with one more task manager:
```
$ oc get pods -n taxi-ride-fare
NAME                                                    READY     STATUS    RESTARTS   AGE
taxi-ride-fare-logger-7fc778d4cf-77prc                  1/1       Running   0          4m
taxi-ride-fare-processor-044b6d96-jm-fbbc89f6f-rjfjn    1/1       Running   0          1m
taxi-ride-fare-processor-044b6d96-tm-6d4fb454cf-742hc   1/1       Running   0          1m
taxi-ride-fare-processor-044b6d96-tm-6d4fb454cf-85z9z   1/1       Running   0          1m
taxi-ride-fare-processor-045d9c13-jm-6bbdb5757c-gtzcq   1/1       Running   0          4m
taxi-ride-fare-processor-045d9c13-tm-d67cc9bcf-8g59b    1/1       Running   0          4m
taxi-ride-fare-taxi-fare-668cc8cc85-bxr6k               1/1       Running   0          4m
taxi-ride-fare-taxi-ride-5f5dd8bcc-t984t                1/1       Running   0          4m
```
Here's the output of `oc describe flinkapp`:
```
Name:         taxi-ride-fare-processor
Namespace:    taxi-ride-fare
Labels:       <none>
Annotations:  <none>
API Version:  flink.k8s.io/v1alpha1
Kind:         FlinkApplication
Metadata:
  Creation Timestamp:  2019-08-06T16:26:36Z
  Finalizers:
    job.finalizers.flink.k8s.io
  Generation:        1
  Resource Version:  26862300
  Self Link:         /apis/flink.k8s.io/v1alpha1/namespaces/taxi-ride-fare/flinkapplications/taxi-ride-fare-processor
  UID:               f5d40450-b866-11e9-8aa5-0aa6034a1df8
Spec:
  Deployment Mode:  Dual
  Entry Class:      pipelines.runner.Runner
  Flink Config:
    State . Backend . Fs . Checkpointdir:  file:///mnt/flink/storage/checkpoints/processor
    State . Checkpoints . Dir:             file:///mnt/flink/storage/externalized-checkpoints/processor
    State . Savepoints . Dir:              file:///mnt/flink/storage/savepoints/processor
  Flink Version:                           1.8
  Image:                                   docker-registry-default.purplehat.lightbend.com/lightbend/taxi-ride-fare@sha256:3d24e4bbaf5327e8ab5223fe72ece1cd8122998b4729d7ae0171d518ef1e2784
  Image Pull Policy:                       Always
  Jar Name:                                runner.jar
  Job Manager Config:
    Env Config:
    Replicas:  1
    Resources:
      Limits:
        Cpu:     2
        Memory:  1Gi
      Requests:
        Cpu:      500m
        Memory:   512Mi
  Parallelism:    4
  Restart Nonce:  
  Savepoint Info:
  Task Manager Config:
    Env Config:
    Resources:
      Limits:
        Cpu:     2
        Memory:  1Gi
      Requests:
        Cpu:     500m
        Memory:  512Mi
    Task Slots:  2
  Volume Mounts:
    Mount Path:  /mnt/flink/storage
    Name:        persistent-storage-vol
    Mount Path:  /etc/pipelines-runner-secret
    Name:        secret-vol
    Mount Path:  /etc/pipelines-runner
    Name:        config-map-vol
  Volumes:
    Config Map:
      Name:  configmap-taxi-ride-fare-processor
    Name:    config-map-vol
    Name:    secret-vol
    Secret:
      Secret Name:  processor
    Name:           persistent-storage-vol
    Persistent Volume Claim:
      Claim Name:  taxi-ride-fare-pvc
Status:
  Cluster Status:
    Available Task Slots:   0
    Healthy Task Managers:  0
  Deploy Hash:              044b6d96
  Failed Update Hash:       
  Job Status:
    Completed Checkpoint Count:  0
    Entry Class:                 pipelines.runner.Runner
    Failed Checkpoint Count:     0
    Health:                      
    Jar Name:                    runner.jar
    Job ID:                      a3c2c7053cb299ae9773ad942a0d91b9
    Job Restart Count:           0
    Last Checkpoint Time:        <nil>
    Last Failing Time:           <nil>
    Parallelism:                 4
    Restore Path:                
    Restore Time:                <nil>
    Start Time:                  <nil>
    State:                       
  Last Updated At:               2019-08-06T16:30:26Z
  Phase:                         Running
Events:
  Type    Reason  Age   From              Message
  ----    ------  ----  ----              -------
  Normal  Create  6m    flinkk8soperator  Flink cluster created
  Normal  Create  5m    flinkk8soperator  Flink job submitted to cluster with id 15558862882b2c98e8c6f368e93821a1
  Normal  Create  2m    flinkk8soperator  Flink cluster created
  Normal  Create  2m    flinkk8soperator  Flink job submitted to cluster with id a3c2c7053cb299ae9773ad942a0d91b9
```

Reply:
Another thing is, doesn't it make more sense to kill the old cluster before the application starts running in the new cluster? Because that will avoid having two copies of the application producing output (and possibly writing duplicate data to some sink) simultaneously.

Reply:
@yuchaoran2011 @debasishg 
I see something weird going on. This should not happen.

What version of the Flink operator are you guys running? Can you try running the latest - `docker.io/lyft/flinkk8soperator:500fe6bd40da8efca4a48bbb1104896be2c1fae8` image ?

The reason we delete the old cluster only after the application starts in the new one is because there can be bug which can prevent jobs from getting starting in the new cluster. In such a case we do a auto Rollback to the old cluster. @yuchaoran2011 Multiple copies of the application should never run at the same time. The job in the second cluster is started only after the job is cancelled in the first cluster.

Just to explain the state transition sequence -
1) New Cluster is created.
2) Wait for the new cluster to bring up all the pods.
3) Once new cluster is up, cancel the Job in the first cluster.
4) Once cancelled, start the job in the second cluster.
5) If the job in the second cluster starts running, delete the old cluster.

Reply:
Hoepfully the state machine diagram makes the sequence clear:
![image](https://user-images.githubusercontent.com/100920/62567745-ae1c0380-b840-11e9-80a4-bcd71f7900d5.png)

The deploy process (on the happy path) is:
1. Create the new cluster
2. Cancel the old job with a savepoint
3. Submit the job to the new cluster
4. Tear down the new cluster

So we're never running two jobs at once, even though both clusters are still running. This sequence was chosen to minimize downtime in the case of a bad deploy. For example, if there's a problem with the new code that prevents it from starting up, we want to immediately be able to resubmit the previous job to the previous cluster without having to wait for it to come back up again.

Reply:
> something is likely going wrong in the DeleteOldResourcesForApp method. Can you see if there's any relevant logs from the operator?

Nothing is there related to this method in the log.

Reply:
Are there any other errors in the log? This could also be caused by an error in `handleApplicationRunning` hitting before the call to `DeleteOldResourcesForApp`

Reply:
I don't see any error in the log ..

Reply:
Tried to use the new tag of the image as mentioned above .. getting an exception starting Task Managers ..

```
2019-08-07 06:16:26,061 ERROR org.apache.flink.runtime.taskexecutor.TaskManagerRunner       - TaskManager initialization failed.
java.lang.Exception: Could not create actor system
	at org.apache.flink.runtime.clusterframework.BootstrapTools.startActorSystem(BootstrapTools.java:267)
	at org.apache.flink.runtime.clusterframework.BootstrapTools.startActorSystem(BootstrapTools.java:153)
	at org.apache.flink.runtime.clusterframework.BootstrapTools.startActorSystem(BootstrapTools.java:112)
	at org.apache.flink.runtime.clusterframework.BootstrapTools.startActorSystem(BootstrapTools.java:87)
	at org.apache.flink.runtime.rpc.akka.AkkaRpcServiceUtils.createRpcService(AkkaRpcServiceUtils.java:84)
	at org.apache.flink.runtime.taskexecutor.TaskManagerRunner.createRpcService(TaskManagerRunner.java:412)
	at org.apache.flink.runtime.taskexecutor.TaskManagerRunner.<init>(TaskManagerRunner.java:134)
	at org.apache.flink.runtime.taskexecutor.TaskManagerRunner.runTaskManager(TaskManagerRunner.java:332)
	at org.apache.flink.runtime.taskexecutor.TaskManagerRunner$1.call(TaskManagerRunner.java:302)
	at org.apache.flink.runtime.taskexecutor.TaskManagerRunner$1.call(TaskManagerRunner.java:299)
	at org.apache.flink.runtime.security.NoOpSecurityContext.runSecured(NoOpSecurityContext.java:30)
	at org.apache.flink.runtime.taskexecutor.TaskManagerRunner.main(TaskManagerRunner.java:299)
Caused by: com.typesafe.config.ConfigException$Parse: String: 55: Expecting a value but got wrong token: 'h' ('$' not followed by {, 'h' not allowed after '$') (if you intended 'h' ('$' not followed by {, 'h' not allowed after '$') to be part of a key or string value, try enclosing the key or value in double quotes, or you may be able to rename the file .properties rather than .conf)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseError(ConfigDocumentParser.java:201)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseError(ConfigDocumentParser.java:197)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseValue(ConfigDocumentParser.java:251)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseObject(ConfigDocumentParser.java:475)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseValue(ConfigDocumentParser.java:247)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseObject(ConfigDocumentParser.java:458)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseValue(ConfigDocumentParser.java:247)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseObject(ConfigDocumentParser.java:458)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseValue(ConfigDocumentParser.java:247)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseObject(ConfigDocumentParser.java:458)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseValue(ConfigDocumentParser.java:247)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parseObject(ConfigDocumentParser.java:458)
	at com.typesafe.config.impl.ConfigDocumentParser$ParseContext.parse(ConfigDocumentParser.java:648)
	at com.typesafe.config.impl.ConfigDocumentParser.parse(ConfigDocumentParser.java:14)
	at com.typesafe.config.impl.Parseable.rawParseValue(Parseable.java:262)
	at com.typesafe.config.impl.Parseable.rawParseValue(Parseable.java:250)
	at com.typesafe.config.impl.Parseable.parseValue(Parseable.java:180)
	at com.typesafe.config.impl.Parseable.parseValue(Parseable.java:174)
	at com.typesafe.config.impl.Parseable.parse(Parseable.java:301)
	at com.typesafe.config.ConfigFactory.parseString(ConfigFactory.java:1051)
	at com.typesafe.config.ConfigFactory.parseString(ConfigFactory.java:1061)
	at org.apache.flink.runtime.akka.AkkaUtils$.getRemoteAkkaConfig(AkkaUtils.scala:601)
	at org.apache.flink.runtime.akka.AkkaUtils$.getAkkaConfig(AkkaUtils.scala:221)
	at org.apache.flink.runtime.akka.AkkaUtils.getAkkaConfig(AkkaUtils.scala)
	at org.apache.flink.runtime.clusterframework.BootstrapTools.startActorSystem(BootstrapTools.java:247)
	... 11 more
```

Incidentally I get the same error with image tag `0.1.3`, which I tried just now. We had been using `0.1.2` which does not give this exception.

By the way we run Flink `1.8` ..

Reply:
You need to update the Docker image. Check this - https://github.com/lyft/flinkk8soperator/commit/373cb21e6d5c7247dfc7c18c7a4d202ce2eedb2a

We recently pushed a backward incompatible change.

Reply:
Yeah .. I changed the image in `flinkk8soperator.yaml` .. do I need to do anything else ?

```
containers:
      - name: flinkoperator-gojson
        image: docker.io/lyft/flinkk8soperator:500fe6bd40da8efca4a48bbb1104896be2c1fae8
```

Reply:
I was going through the change .. in lines `186` and `192` of `pkg/controller/flink/container_utils.go` ..

did u mean `$HOST_IP` or should they be `${?HOST_IP}` if they are meant  to be picked up for processing by Akka config in Flink ?

My hunch was purely based on the error that we face where it complains about `$` being followed by `h` without a `{` ..

Reply:
I am not referring to the flink operator image. This is the application image. 

You will need to update your application image to be able to run with the latedt version of the operator. Check the changes here - https://github.com/lyft/flinkk8soperator/commit/373cb21e6d5c7247dfc7c18c7a4d202ce2eedb2a